src/trs/lexicon.rs

```diff
@@ -805,7 +838,46 @@ impl Lex {
         }
         context.to_rule().or(Err(SampleError::Subterm))
     }
+    fn infer_term_inter(
+        &self,
+        term: &Term,
+        ctx: &mut TypeContext,
+        place: Place,
+        tps: &mut HashMap<Place, Type>,
+    ) -> Result<Type, TypeError> {
+        let tp = match *term {
+            Term::Variable(ref v) => self.instantiate_atom(&Atom::from(v.clone()), ctx)?,
+            Term::Application { ref op, ref args } => {
+                let head_type = self.instantiate_atom(&Atom::from(op.clone()), ctx)?;
+                let body_type = {
+                    let mut pre_types = Vec::with_capacity(args.len() + 1);
+                    for (i, a) in args.iter().enumerate() {
+                        let mut new_place = place.clone();
+                        new_place.push(i);
+                        pre_types.push(self.infer_term_inter(a, ctx, new_place, tps)?);
+                    }
+                    pre_types.push(ctx.new_variable());
+                    Type::from(pre_types)
+                };
+                ctx.unify(&head_type, &body_type)?;
+                head_type.returns().unwrap_or(&head_type).apply(ctx)
+            }
+        };
+        tps.insert(place, tp.clone());
+        Ok(tp)
+    }
 
+    fn var_type_to_replace_common_term(
```

Comment:
Can you please add some documentation? I'm not entirely sure what this function does?